// @creit.tech/stellar-wallets-kit/modules/freighter.module.cjs
'use strict';

var freighterApi = require('@stellar/freighter-api');
var types = require('../types.cjs');

const FREIGHTER_ID = "freighter";
class FreighterModule {
  constructor() {
    this.moduleType = types.ModuleType.HOT_WALLET;
    this.productId = FREIGHTER_ID;
    this.productName = "Freighter";
    this.productUrl = "https://freighter.app";
    this.productIcon = "https://stellar.creit.tech/wallet-icons/freighter.svg";
  }
  async isAvailable() {
    return freighterApi.isConnected();
  }
  async getPublicKey() {
    if (!await freighterApi.isConnected()) {
      throw new Error(`Freighter is not connected`);
    }
    return freighterApi.requestAccess();
  }
  async signTx(params) {
    if (!await freighterApi.isConnected()) {
      throw new Error(`Freighter is not connected`);
    }
    let updatedXdr = params.xdr;
    for (const publicKey of params.publicKeys) {
      updatedXdr = await freighterApi.signTransaction(updatedXdr, {
        accountToSign: publicKey,
        networkPassphrase: params.network
      });
    }
    return { result: updatedXdr };
  }
  async signBlob(params) {
    if (!await freighterApi.isConnected()) {
      throw new Error(`Freighter is not connected`);
    }
    const result = await freighterApi.signBlob(params.blob, { accountToSign: params.publicKey });
    return { result };
  }
  async signAuthEntry(params) {
    if (!await freighterApi.isConnected()) {
      throw new Error(`Freighter is not connected`);
    }
    const result = await freighterApi.signAuthEntry(params.entryPreimageXDR, {
      accountToSign: params.publicKey
    });
    return { result };
  }
}

exports.FREIGHTER_ID = FREIGHTER_ID;
exports.FreighterModule = FreighterModule;
