// @creit.tech/stellar-wallets-kit/modules/lobstr.module.cjs
'use strict';

var signerExtensionApi = require('@lobstrco/signer-extension-api');
var types = require('../types.cjs');

const LOBSTR_ID = "lobstr";
class LobstrModule {
  constructor() {
    this.moduleType = types.ModuleType.HOT_WALLET;
    this.productId = LOBSTR_ID;
    this.productName = "LOBSTR";
    this.productUrl = "https://lobstr.co";
    this.productIcon = "https://stellar.creit.tech/wallet-icons/lobstr.svg";
  }
  async isAvailable() {
    return signerExtensionApi.isConnected();
  }
  async getPublicKey() {
    if (!await signerExtensionApi.isConnected()) {
      throw new Error(`Lobstr is not connected`);
    }
    return signerExtensionApi.getPublicKey();
  }
  async signTx(params) {
    if (!await signerExtensionApi.isConnected()) {
      throw new Error(`Lobstr is not connected`);
    }
    if (params.publicKeys.length > 0) {
      console.warn(`Lobstr doesn't allow specifying what public key should sign the transaction, we skip the value`);
    }
    if (params.network) {
      console.warn(`Lobstr doesn't allow specifying the network that should be used, we skip the value`);
    }
    return { result: await signerExtensionApi.signTransaction(params.xdr) };
  }
  // @ts-expect-error - This is not a supported operation so we don't use the params
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async signBlob(params) {
    throw new Error("Lobstr does not support signing random blobs");
  }
  // @ts-expect-error - This is not a supported operation so we don't use the params
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async signAuthEntry(params) {
    throw new Error("Lobstr does not support signing authorization entries");
  }
}

exports.LOBSTR_ID = LOBSTR_ID;
exports.LobstrModule = LobstrModule;
