// @creit.tech/stellar-wallets-kit/index.cjs
'use strict';

var stellarWalletsKit = require('./stellar-wallets-kit.cjs');
var stellarWalletsModal = require('./modal/stellar-wallets-modal.cjs');
var types = require('./types.cjs');
var utils = require('./utils.cjs');
var xbull_module = require('./modules/xbull.module.cjs');
var freighter_module = require('./modules/freighter.module.cjs');
var albedo_module = require('./modules/albedo.module.cjs');
var rabet_module = require('./modules/rabet.module.cjs');
var lobstr_module = require('./modules/lobstr.module.cjs');
var hana_module = require('./modules/hana.module.cjs');



exports.StellarWalletsKit = stellarWalletsKit.StellarWalletsKit;
Object.defineProperty(exports, "StellarWalletsModal", {
	enumerable: true,
	get: function () { return stellarWalletsModal.StellarWalletsModal; }
});
exports.ModuleType = types.ModuleType;
exports.WalletNetwork = types.WalletNetwork;
exports.allowAllModules = utils.allowAllModules;
exports.XBULL_ID = xbull_module.XBULL_ID;
exports.xBullModule = xbull_module.xBullModule;
exports.FREIGHTER_ID = freighter_module.FREIGHTER_ID;
exports.FreighterModule = freighter_module.FreighterModule;
exports.ALBEDO_ID = albedo_module.ALBEDO_ID;
exports.AlbedoModule = albedo_module.AlbedoModule;
exports.AlbedoNetwork = albedo_module.AlbedoNetwork;
exports.RABET_ID = rabet_module.RABET_ID;
exports.RabetModule = rabet_module.RabetModule;
exports.RabetNetwork = rabet_module.RabetNetwork;
exports.LOBSTR_ID = lobstr_module.LOBSTR_ID;
exports.LobstrModule = lobstr_module.LobstrModule;
exports.HANA_ID = hana_module.HANA_ID;
exports.HanaModule = hana_module.HanaModule;
